### ISSUE TITLE
Hide iOS blinking blue cursor *without* breaking typing in Chrome

### ISSUE THREAD

**Issue by laughinghan** [mathquill/mathquill#584]
So maybe setting `font-size: 0` on the textarea to hide the blinking blue cursor on iOS was not so ingenious after all (323724b20c5a18eec1147d23b2cc65f13686f930 on #452), it's causing:
- all keyboard input to be ignored in Chrome Canary 51 (#540, which I've repro'd and confirmed goes away without the `font-size: 0`, and reported to [Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=595943)),
- the Enter key to be ignored in Chrome 48 and Safari 9 (MathQuill doesn't let you type newlines but is supposed to fire an `enter` handler: #566; reported to [Webkit](https://bugs.webkit.org/show_bug.cgi?id=155615#c0) and mentioned in the Chromium issue),
- in Safari on iOS 9.2.1 (and probably other mobile browsers too), focusing a math field zooms _wayyyy_ in (happens on the http://mathquill.com homepage and local test page /tests/demo.html),
- and probably the other typing issues as well (#559).

Okay so `font-size: 0` is not going to cut it. In that case, how do we hide the blinking blue cursor in iOS? Options:

- position the hidden textarea past the left (or top) edge of the viewport. This is a perennially popular solution, it's been suggested in this project a few times (9b323e3347ea4be55fd77e75f17a37d08889f9dc, 07578edd7537f1000005b55f0b143acc696223c3) and is the only suggestion that works for iOS and IE (and not just Gecko and Webkit like used to be the case for MathQuill's hidden textarea (1fa701b84b0e53e46bebb9628ad1ceeb04b08fc3)) that I've seen on StackOverflow (e.g. [this elegantly hand-drawn diagram](http://stackoverflow.com/a/30010698/362030), or [here](http://stackoverflow.com/a/29146606/362030) or [here](http://stackoverflow.com/a/30950258/362030))
    + Since this is clearly what most of the Internet is doing, it's a good idea to have this in our back pocket.
    + However, this technique has a major caveat: when you focus a textarea, all browsers scroll it into view, and mobile browsers zoom into it. This doesn't matter to full-page webapps that can't scroll at all like the Desmos graphing calculator, but if your page or app can scroll and our hidden textarea is past the top edge of the viewport, we'll scroll to the top whenever MathQuill gets focused. Which is why many implementations of this technique put the hidden textarea past the left edge of the viewport without affecting it's vertical coordinate, this way desktop browsers usually scroll to exactly the right place, as do mobile browsers on pages with [`user-scalable=no`](https://developer.apple.com/library/iad/documentation/AppleApplications/Reference/SafariHTMLRef/Articles/MetaTags.html#//apple_ref/doc/uid/TP40008193-SW6). However, this is still problematic on normal pages in mobile browsers, because this means that when focusing a MathQuill math field, they'll zoom into the left edge of the screen even if the math field is in the center of the screen (similarly, on a page wide enough to need horizontal scrolling, desktop browsers will scroll all the way to the left trying to scroll-into-view the textarea; but pages needing horizontal scrolling in a desktop browser are pretty uncommon).
- browser-sniff for iOS (or mobile?) and selectively apply `line-height: 0` to the textarea (which has the same desired behavior of hiding the blinking blue cursor on iOS and undesired behavior of breaking typing sometimes; but doesn't have the issue where Safari on iOS zooms _wayyyy_ in when focused)
    + This feels like admitting defeat. We avoided browser sniffing when none of the straightforward ways to vertically align the fraction bar correctly always failed in at least one of Gecko, Webkit, and Opera; we avoided browser sniffing when keyboard events were ["madness"](http://unixpapa.com/js/key.html); we're gonna be defeated by a goddamn blinking cursor?
- I did find one new technique that I hadn't discovered during #452, injecting a newline at the beginning of a textarea that's less than one line and scrolling to the top, so that the cursor is always out of view: http://jsbin.com/cohema/edit?css,js,output
    + I really do appear to _have to_ inject a newline, [my other attempts](https://gist.github.com/laughinghan/faba0e276a5c3ea30088) to insert space and scroll the cursor out of view all result in the textarea increasing in height/width until a bit of the blinking blue cursor "peeks out".
    + Unfortunately injecting a newline means this probably isn't a viable near-term solution, since that interferes with us figuring out what people typed by [reading the contents of the textarea](https://github.com/mathquill/mathquill/blob/v0.10.0/src/services/saneKeyboardEvents.util.js#L191) (that's the whole point of the hidden textarea in the first place); we would have to change our already disgustingly complex keyboard events cross-browser madness-queller to ignore the injected newline, without ignoring a real newline that gets typed (#566 is what started me on this in the first place).

Out of all of these, the best option looks like browser-sniffing :disappointed: . Unlike the inject-newline-and-scroll-down thing, it's plug-and-play, we don't have to futz with the keyboard events code; and it has a theoretical potential future downside, but the positioning the textarea past the edge of the viewport has a concrete downside today.

**Comment by laughinghan** [comment]
EXCEPT NEVERMIND FORGET EVERYTHING 'COS I'M A GENIUS CHECK THIS OUT: http://jsbin.com/sodaji/edit?html,css,js,output

God CSS transforms are so amazing why weren't they the first thing I tried?!

Only tested in Safari on iOS 9.2.1 but unlike `font-size: 0` which let's be honest always felt kinda uncomfortable, this feels like it's probably actually super robust. Transforms are understood to happen at a really low level, after everything, in shitty browsers it can be after *rasterizing*; transforms are divorced from CSS and reflow and the DOM and all that to a much greater extent than `position: relative` and stuff.

I honestly feel so good about the compatibility of `transform: scale(0)` (and browser-prefixed equivalents) that I'm expecting it to work in IE, which means we can even get rid of the `clip: rect()` business.

**Comment by fnlctrl** [comment]
Thanks for the awesome `transform: scale(0)` solution!

It could be better though, if we don't need to manually bind click events to focus that textarea/input.
So I came up with this: I put a `<label>` beside the `<textarea>` and generated an id for them, and make them `position:absolute` to overlay on each other. So when `<label>` is clicked, the `<textarea>` is automatically focused, thanks to the browser.

**Comment by laughinghan** [comment]
@fnlctrl Cute, what are you using this technique for? A caveat of the label technique is that it doesn't focus the textarea until the `click` event fires ([try it](https://output.jsbin.com/lafevoz)), whereas real textareas focus as soon as you `mousedown`, and so since MathQuill is pretending to be a textarea but for math, it needs to manually bind a `mousedown` event handler to immediately focus the hidden textarea.

**Comment by laughinghan** [comment]
For posterity, Desmos ran into a crazy/amazing situation related to these browser behaviors (bugs, arguably): setting `font-size: 1px` on a textarea doesn't break typing in Chrome 51, but if you zoom out to a 90% level, typing will then be broken. This was an absolute nightmare to track down because people encountering the problem didn't even realize they were zoomed out, and someone else running the same code in the same browser version on the same OS version would be unable to reproduce if they were at normal zoom. What seemed extra bizarre was that though the person wouldn't encounter the problem in another browser, they would in any other tab or window, _even an Incognito window_, even after restarting the browser or machine, but visiting the same code in the same browser _during the same session_ but at a different domain name wouldn't exhibit the problem, all because Chrome helpfully persists zoom information and shares it across sessions on a per-domain basis.

TL;DR new lesson: in addition to asking for bug reporter's browser and OS version, also ask for zoom level.
